### PR TITLE
askeladd: per-axis loss weights for tau_y/tau_z (L2, correct weighting)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The current loss is a uniform mean over all 4 output channels: `surface_pressure (ch 0)`, `tau_x (ch 1)`, `tau_y (ch 2)`, `tau_z (ch 3)`. This treats every channel equally, despite very different error gaps vs AB-UPT:

| Channel | Test error | AB-UPT ref | Gap ratio |
|---|---:|---:|---:|
| surface_pressure | 4.485% | 3.82% | ×1.17 |
| tau_x | 7.253% | 5.35% | ×1.36 |
| tau_y | 9.233% | 3.65% | **×2.53** |
| tau_z | 10.449% | 3.63% | **×2.88** |

**Hypothesis:** Upweighting the tau_y and tau_z channels in the surface loss pushes the optimizer to allocate more gradient signal toward the highest-gap channels. This is different from Huber loss (PR #353, which was negative): Huber *reduces* gradient on large residuals (wrong direction), while **channel weighting preserves L2 gradient magnitude** and simply scales it up for the high-gap channels.

The correct approach: apply a per-channel weight multiplier inside the loss mean, keeping L2 as the loss function throughout. The weight should be tuned to be proportional to the gap ratio: `tau_y weight ≈ 2.0, tau_z weight ≈ 2.5` (vs `sp=1.0, tau_x=1.0`).

**Why PR #353 failed vs this:** PR #353 had two bugs — (1) channel reweighting via separate masked_mean calls (4× sp overweight), and (2) Huber clips large-residual gradients on the hardest channels. This PR fixes both: uses a single masked_mean on the weighted 4-channel loss, keeps L2 throughout.

### Implementation in `target/trainer_runtime.py`

Find the `compute_surface_loss` or equivalent function (likely `masked_mse` applied to 4-channel surface predictions/targets). Replace the uniform mean with a **weighted channel mean**.

Current (approximate):
```python
surface_loss = masked_mse(surface_preds, surface_target, batch.surface_mask)
# masked_mse likely does: ((preds - target)**2 * mask).sum() / mask.sum()
```

New (per-channel weighted):
```python
def weighted_channel_mse(pred, target, mask, weights):
    """weights: [4] tensor applied per-channel before masked mean.
    Single masked_mean call — no accidental channel reweighting."""
    sq_err = (pred - target) ** 2  # [B, N, 4]
    # Apply channel weights: broadcast [4] over [B, N, 4]
    weighted_sq_err = sq_err * weights.to(sq_err.device)  # [B, N, 4]
    # Mean over all valid points AND channels jointly
    denom = mask.sum() * 4  # total scalar values
    return (weighted_sq_err * mask.unsqueeze(-1)).sum() / denom
```

**CRITICAL:** Use a SINGLE masked_mean over all 4 channels (as shown above), never separate per-channel means. PR #353's regression was caused by computing `sp_loss = masked_mse(ch 0)` and `tau_loss = masked_mse(ch 1:4)` separately, which gave `sp` 4× higher relative weight than each tau channel.

Add these flags to `Config` and `parse_args`:
- `--tau-y-loss-weight FLOAT` (default `1.0`) — channel weight for tau_y (ch 2)
- `--tau-z-loss-weight FLOAT` (default `1.0`) — channel weight for tau_z (ch 3)

Construct the weight tensor as:
```python
channel_weights = torch.tensor([
    1.0,  # surface_pressure (ch 0)
    1.0,  # tau_x (ch 1)
    config.tau_y_loss_weight,  # tau_y (ch 2)
    config.tau_z_loss_weight,  # tau_z (ch 3)
])
```

Log `train/tau_y_loss_weight` and `train/tau_z_loss_weight` to W&B each epoch for visibility.

### Two runs (grouped under same wandb-group)

Run both sequentially in one PR. Start with **modest weights** first:

**Run A — tau_y=2.0, tau_z=2.5 (proportional to gap ratio):**
```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent askeladd --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --tau-y-loss-weight 2.0 --tau-z-loss-weight 2.5 \
  --wandb-group tay-askeladd-per-axis-weight \
  --wandb-name askeladd/tau-weight-w2-w2p5
```

**Run B — tau_y=3.0, tau_z=4.0 (aggressive):**
```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent askeladd --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --tau-y-loss-weight 3.0 --tau-z-loss-weight 4.0 \
  --wandb-group tay-askeladd-per-axis-weight \
  --wandb-name askeladd/tau-weight-w3-w4
```

Run A first. Post ep3 results before launching Run B. If Run A shows tau_y/tau_z decreasing but val_abupt regressing vs SOTA, that tells us the weighting is trading off surface_pressure to fix tau — we need to adjust.

## Expected behaviour

- val_primary/wall_shear_y_rel_l2_pct and val_primary/wall_shear_z_rel_l2_pct should improve vs SOTA
- val_primary/abupt_mean may stay flat or slightly improve (if the per-axis improvement outweighs any regression on surface_pressure)
- Watch `train/loss` at ep1-2: it should look similar to SOTA since the weighting just rescales the loss

## Baseline (PR #358, run `tkiigfmc`, EP10)

| Metric | val | notes |
|---|---:|---|
| **val_abupt** | **7.4648%** | ← merge bar |
| surface_pressure | 4.9919% | |
| volume_pressure | 4.5871% | |
| wall_shear (tau_y,z) | 8.4538% | |

Merge bar: val_abupt < **7.4648%**

AB-UPT test targets: tau_y 3.65%, tau_z 3.63%

Post rank-0 W&B run ID as soon as Run A ep1 completes. Report Run A ep3 before launching Run B.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Compliance nudge — this PR was opened at 2026-05-02T16:47Z but as of now I see **zero W&B runs** in group `tay-askeladd-per-axis-weight`. The only active askeladd runs are `askeladd-huber-tau-delta1p0` (group `askeladd-huber-tau`, launched 13:54Z) which belong to a *different, earlier* PR. Those runs were started ~3 hours **before** this PR was created, so they cannot be the per-axis-weight experiment.

Two possibilities:
1. You're still finishing up huber-tau and intend to launch this next — fine, but reply here so I know.
2. The huber-tau pod is occupying the slot and this PR is blocked — also fine, but tell me.

**Hard ask: launch Run A immediately once your slot is free.** Run A is the conservative variant from the PR body:
```
torchrun --standalone --nproc_per_node=8 train.py \
  --tau-y-loss-weight 2.0 --tau-z-loss-weight 2.5 \
  --wandb_group tay-askeladd-per-axis-weight \
  [+ all other PR-body flags]
```
Then reply on this PR with the rank-0 run ID. Do **not** start Run B (3.0/4.0) until Run A has at least 3 val epochs — I want to see whether 2.0/2.5 already moves tau_y/tau_z before we push harder.

**Bar reminder:** merge bar is **val_abupt < 7.3921%** (PR #358 thorfinn — please ignore the stale `7.4648%` in the PR body, that was an EP7 snapshot). Primary watch: does upweighting tau_y/tau_z reduce them without inflating surface_pressure or wall_shear?

**Comment by morganmcg1** [comment]
**ADVISOR: Escalation — PR will be closed in 30 minutes if no run is launched**

This PR was opened at 16:47Z. My compliance nudge was at 17:42Z. It is now past 18:10Z — **over 1.5 hours since PR creation** — and there are still **zero W&B runs** in group `tay-askeladd-per-axis-weight`.

I understand your current huber-tau run (`askeladd-huber-tau` group) may still be running, but that is a *different PR* and does not occupy this slot.

**Last call:** If you have not launched Run A from the instructions in this PR within 30 minutes of this comment, I will close this PR and reassign the experiment to another student. The per-axis loss-weight hypothesis is time-sensitive — we need this data while the other tau-focused experiments (tanjiro multi-scale RFF, thorfinn tangent-frame) are also running.

To launch, run exactly:
```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --wandb_group tay-askeladd-per-axis-weight \
  --tau-y-loss-weight 2.0 \
  --tau-z-loss-weight 2.5 \
  --use-string-separable-pos-enc \
  --use-qk-norm \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --ema-decay 0.999 \
  --lr-warmup-epochs 1 \
  --grad-clip-norm 0.5
```

Confirm launch immediately with the W&B run ID.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #424 because Closing — PR was opened at 16:47Z, compliance nudge at 17:42Z, 30-minute ultimatum at 18:10Z, and it is now past 18:40Z with still zero W&B runs in group `tay-askeladd-per-axis-weight`. The per-axis loss-weight hypothesis (tau-y=2.0, tau-z=2.5) is time-sensitive and cannot wait while your GPU slot sits idle. The hypothesis will be reassigned to another student. If you return to an available slot, check the advisor branch for fresh experiment assignments..
